Net/SMTP.php

```diff
@@ -1109,20 +1126,54 @@ protected function authGSSAPI($uid, $pwd, $authz = '')
      *               kind of failure, or true on success.
      * @since 1.9.0
      */
-    public function authXOAuth2($uid, $token, $authz, $conn)
+    protected function authXOAuth2($uid, $token, $authz, $conn)
     {
         $auth = base64_encode("user=$uid\1auth=$token\1\1");
+        return $this->authenticateOAuth('XOAUTH2', $auth, $authz, $conn);
+    }
+
+    /**
+     * Authenticates the user using the OAUTHBEARER method.
+     *
+     * @param string $uid   The userid to authenticate as.
+     * @param string $token The access token to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     * @param object $conn  The current object
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since 1.9.3
+     * @see https://www.rfc-editor.org/rfc/rfc7628.html
+     */
+    protected function authOAuthBearer($uid, $token, $authz, $conn)
+    {
+        $auth = base64_encode("n,a=$uid\1auth=$token\1\1");
```

Comment:
Yes it is implemented on my Roundcube's instance (https://github.com/roundcube/roundcubemail/pull/9289/files)  and successfully working with my Postfix

I kept the same logic from the previous `authXOAuth2()` where the token is prefixed by it's type (`Bearer` in this case)
https://developers.google.com/gmail/imap/xoauth2-protocol?hl=fr for XOauth2

The authorization payload is generated here in Roundcube:
https://github.com/roundcube/roundcubemail/blob/357cc90001f997fd223fb48fcede6040f527c2f4/program/include/rcmail_oauth.php#L911

So if we want to be more precise I suggest to correct auth*OAuth() and rename `$token` into `$auth`

Reply:
Hello @schengawegga @thomascube 

regarding the issue with `auth=Bearer $token` versus  the version that has been previously implemented in `authXOAuth()`: `auth=$auth`

Here is options I see:

### option 1: use the real token and generate the payload adding 'Bearer' prefix

pro: 
- cleaner and same logic with implementation on other languages

cons:
-  will **break compatibility** with the previous implementation of `authXOAuth2()`
-  requires a change in Roundcube & Net_Sieve (not a problem I can submit a patch on both)

```diff
--- a/Net/SMTP.php
+++ b/Net/SMTP.php
@@ -1128,7 +1128,7 @@ class Net_SMTP
      */
     protected function authXOAuth2($uid, $token, $authz, $conn)
     {
-        $auth = base64_encode("user=$uid\1auth=$token\1\1");
+        $auth = base64_encode("user=$uid\1auth=Bearer $token\1\1");
         return $this->authenticateOAuth('XOAUTH2', $auth, $authz, $conn);
     }

@@ -1147,7 +1147,7 @@ class Net_SMTP
      */
     protected function authOAuthBearer($uid, $token, $authz, $conn)
     {
-        $auth = base64_encode("n,a=$uid\1auth=$token\1\1");
+        $auth = base64_encode("n,a=$uid\1auth=Bearer $token\1\1");
         return $this->authenticateOAuth('OAUTHBEARER', $auth, $authz, $conn);
     }
```

### option 2: change the naming to be explicit and avoid any confusion

pro: no change on behavior, just a change on variable name to avoid confusion
cons: code that will use the library must prefix explicitlythe `Bearer `
```diff
index efb1745..26c429e 100644
--- a/Net/SMTP.php
+++ b/Net/SMTP.php
@@ -1118,7 +1118,7 @@ class Net_SMTP
      * Authenticates the user using the XOAUTH2 method.
      *
      * @param string $uid   The userid to authenticate as.
-     * @param string $token The access token to authenticate with.
+     * @param string $auth  The authentification (format: "Bearer $access_token")
      * @param string $authz The optional authorization proxy identifier.
      * @param object $conn  The current object
      *
@@ -1126,17 +1126,17 @@ class Net_SMTP
      *               kind of failure, or true on success.
      * @since 1.9.0
      */
-    protected function authXOAuth2($uid, $token, $authz, $conn)
+    protected function authXOAuth2($uid, $auth, $authz, $conn)
     {
-        $auth = base64_encode("user=$uid\1auth=$token\1\1");
-        return $this->authenticateOAuth('XOAUTH2', $auth, $authz, $conn);
+        $payload = base64_encode("user=$uid\1auth=$auth\1\1");
+        return $this->authenticateOAuth('XOAUTH2', $payload, $authz, $conn);
     }

     /**
      * Authenticates the user using the OAUTHBEARER method.
      *
      * @param string $uid   The userid to authenticate as.
-     * @param string $token The access token to authenticate with.
+     * @param string $auth  The authentification (format: "Bearer $access_token")
      * @param string $authz The optional authorization proxy identifier.
      * @param object $conn  The current object
      *
@@ -1145,31 +1145,31 @@ class Net_SMTP
      * @since 1.9.3
      * @see https://www.rfc-editor.org/rfc/rfc7628.html
      */
-    protected function authOAuthBearer($uid, $token, $authz, $conn)
+    protected function authOAuthBearer($uid, $auth, $authz, $conn)
     {
-        $auth = base64_encode("n,a=$uid\1auth=$token\1\1");
-        return $this->authenticateOAuth('OAUTHBEARER', $auth, $authz, $conn);
+        $payload = base64_encode("n,a=$uid\1auth=$auth\1\1");
+        return $this->authenticateOAuth('OAUTHBEARER', $payload, $authz, $conn);
     }

     /**
      * Authenticates the user using the OAUTHBEARER or XOAUTH2 method.
      *
      * @param string $method  The method (OAUTHBEARER or XOAUTH2)
-     * @param string $auth   The authentication string (base64 coded)
+     * @param string $payload The OAuthBearer or XOAuth2 payload
      * @param string $authz   The optional authorization proxy identifier.
      * @param object $conn    The current object
      *
      * @return mixed Returns a PEAR_Error with an error message on any
      *               kind of failure, or true on success.
      */
-    protected function authenticateOAuth( $method, $auth, $authz, $conn)
+    protected function authenticateOAuth( $method, $payload, $payloadz, $conn)
     {
         // Maximum length of the base64-encoded token to be sent in the initial response is 504 - strlen($method) bytes,
         // according to RFC 4954 (https://datatracker.ietf.org/doc/html/rfc4954); for longer tokens an empty initial
         // response MUST be sent and the token must be sent separately
         // (504 bytes = /SMTP command length limit/ - 6 bytes /"AUTH "/ -strlen($method) - 1 byte /" "/ - 2 bytes /CRLF/)
-        if (strlen($auth) <= (504-strlen($method))) {
-            if (PEAR::isError($error = $this->put('AUTH', $method . ' ' . $auth))) {
+        if (strlen($payload) <= (504-strlen($method))) {
+            if (PEAR::isError($error = $this->put('AUTH', $method . ' ' . $payload))) {
                 return $error;
             }
         } else {
@@ -1183,7 +1183,7 @@ class Net_SMTP
             }

             // then follows the token
-            if (PEAR::isError($error = $this->put($auth))) {
+            if (PEAR::isError($error = $this->put($payload))) {
                 return $error;
             }
         } 
```

### option3: hybrid: automatically normalize
pro: no change and permit the cleaner syntax in option1, useful for a transitional phase to option1
cons: not fan of "automagic" solutions 
```diff
diff --git a/Net/SMTP.php b/Net/SMTP.php
index efb1745..1d7ea0d 100644
--- a/Net/SMTP.php
+++ b/Net/SMTP.php
@@ -1128,6 +1128,10 @@ class Net_SMTP
      */
     protected function authXOAuth2($uid, $token, $authz, $conn)
     {
+        // prefix the token type Bearer (if not already present)
+        if (strncasecmp($token, 'Bearer ', 7) != 0) {
+            $token = "Bearer $token";
+        }
         $auth = base64_encode("user=$uid\1auth=$token\1\1");
         return $this->authenticateOAuth('XOAUTH2', $auth, $authz, $conn);
     }
@@ -1147,6 +1151,10 @@ class Net_SMTP
      */
     protected function authOAuthBearer($uid, $token, $authz, $conn)
     {
+        // prefix the token type Bearer (if not already present)
+        if (strncasecmp($token, 'Bearer ', 7) != 0) {
+            $token = "Bearer $token";
+        }
         $auth = base64_encode("n,a=$uid\1auth=$token\1\1");
         return $this->authenticateOAuth('OAUTHBEARER', $auth, $authz, $conn);
     }
```

Do you have any preferences ?
Personally I would recommend option3 even if I am not fan

Reply:
@EdouardVanbelle Thanks for your suggestions.
i would prefer option2, but only the change in the description
```php
-     * @param string $token The access token to authenticate with.
+     * @param string $auth  The authentification (format: "Bearer $access_token")
```
because, this explaines the use of this method better with no compatibility problems and most control and flexibility for the users.

I think renaming the parameters are not necessary, but on this, i do not have a strong opinion.

Option 1 is not good, because of breaking the compatibility and option 3 is not my favourite because of changing the behavior due automatic restriction.